text/0000-alloc-me-maybe.md

```diff
@@ -0,0 +1,547 @@
+
+Some embedded systems do use unwinding, but this is very rare, so it cannot be assumed.
+
+It seems they would be happy to have some system to prevent infallible allocations from ever being used.
+
+
+
+
+## User Profile: Gecko
+
+Gecko is also primarily well-aligned with Rust's current strategy. For the most part, they liberally allocate and are happy to crash on OOM. This is especially palatable now that firefox is multiprocess. However as a *quality of implementation* matter, they occasionally make some subroutines fallible. For instance, it would be unfortunate if a single giant image prevented a page from loading. Similarly, running out of memory while processing a style sheet isn't significantly different from failing to download it.
+
+However in contrast to the embedded case, this isn't done in a particularly principled way. Some parts might be fallible, some might be infallible. Nothing is pooled to isolate tasks. It's just a best-effort affair.
+
+Gecko is built without unwinding.
+
+It seems they would be happy to have some system to prevent infallible allocations from ever being used.
+
+Gecko's need for this API as soon as possible will result in it temporarily forking several of the std collections, which is the primary impetus for this RFC.
+
+
+
+
+
+## User Profile: Server
+
+This represents a commodity server which handles tasks using threads or futures.
+
+Similar to the embedded case, handling allocation failure at the granularity of tasks is ideal for quality-of-implementation purposes. However, unlike embedded development, it isn't considered practical (in terms of cost) to properly take control of everything and ensure allocation failure is handled robustly.
+
+Here unwinding is available, and seems to be the preferred solution, as it maximizes the chances of allocation failures bubbling out of whatever libraries are used. This is unlikely to be totally robust, but that's ok.
+
+With unwinding there isn't any apparent use for an infallible allocation checker.
+
+
+
+
+
+## User Profile: Runtime
+
+A garbage-collected runtime (such as SpiderMonkey or the Microsoft CLR), is generally expected to avoid crashing due to out-of-memory conditions. Different strategies and allocators are used for different situations here. Most notably, there are allocations on the GC heap for the running script, and allocations on the global heap for the actual runtime's own processing (e.g. performing a JIT compilation).
```

Comment:
Yes, I should be more clear. I mostly meant crashing due to runtime internals, but script stuff should also try to recover (e.g. triggering a GC when a script allocation fails and retrying). I ended up cutting all focus from the script side because, as I note just below, the script allocations aren't actually relevant to this RFC (AFAICT).

I didn't include the JVM because I had only found the time to interviewed SM and CLR people.